File: backend/routers/files.py
"""
FILE: backend/routers/files.py
File upload, parsing, and AI analysis.
"""

import os
import uuid
import shutil
from fastapi import APIRouter, UploadFile, File, Form, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select

from database import get_db
from models import UploadedFile
from services.file_parser import parse_file
from services.cloudflare import auto_select_model, chat_complete
from config import get_settings

router = APIRouter()
settings = get_settings()

ALLOWED_EXTENSIONS = {
    "pdf", "docx", "doc", "xlsx", "xls", "csv",
    "png", "jpg", "jpeg", "gif", "webp", "bmp",
    "mp3", "mp4", "wav", "m4a", "ogg", "webm",
    "txt", "md", "json", "xml", "html", "py", "js", "ts",
}


@router.post("/upload")
async def upload_file(
    file: UploadFile = File(...),
    question: str = Form("Analyze this file and summarize its key contents."),
    db: AsyncSession = Depends(get_db),
):
    """Upload a file and immediately analyze it with AI."""
    ext = file.filename.split(".")[-1].lower() if "." in file.filename else ""
    if ext not in ALLOWED_EXTENSIONS:
        raise HTTPException(status_code=400, detail=f"File type .{ext} not supported.")

    # Save file
    file_id = str(uuid.uuid4())
    filename = f"{file_id}.{ext}"
    filepath = os.path.join(settings.upload_dir, filename)
    os.makedirs(settings.upload_dir, exist_ok=True)

    with open(filepath, "wb") as f:
        shutil.copyfileobj(file.file, f)

    file_size = os.path.getsize(filepath)

    # Parse content
    parsed = parse_file(filepath, ext)

    # Pick model
    model_key = auto_select_model(question, has_file=True, file_type=ext)

    # Build context for AI
    file_context = f"""[Uploaded file: {file.filename}]
Type: {ext.upper()} | Size: {file_size:,} bytes

Extracted content:
{parsed['text'][:6000]}
"""
    if parsed.get("tables"):
        file_context += f"\n[Tables found: {len(parsed['tables'])}]"
    if parsed.get("metadata"):
        file_context += f"\n[Metadata: {parsed['metadata']}]"

    messages = [
        {"role": "user", "content": f"{file_context}\n\nUser question: {question}"}
    ]

    result = await chat_complete(messages=messages, model_key=model_key)

    # Save to DB
    db_file = UploadedFile(
        id=file_id,
        filename=filename,
        original_name=file.filename,
        file_type=ext,
        file_size=file_size,
        content_summary=result["text"][:1000],
        is_knowledge_base=False,
    )
    db.add(db_file)
    await db.commit()

    return {
        "file_id": file_id,
        "filename": file.filename,
        "file_type": ext,
        "file_size": file_size,
        "parsed_text_length": len(parsed["text"]),
        "tables_found": len(parsed.get("tables", [])),
        "analysis": result["text"],
        "model_used": model_key,
    }


@router.get("/")
async def list_files(db: AsyncSession = Depends(get_db)):
    result = await db.execute(
        select(UploadedFile).order_by(UploadedFile.created_at.desc()).limit(100)
    )
    files = result.scalars().all()
    return [
        {
            "id": f.id,
            "original_name": f.original_name,
            "file_type": f.file_type,
            "file_size": f.file_size,
            "content_summary": f.content_summary,
            "is_knowledge_base": f.is_knowledge_base,
            "created_at": f.created_at.isoformat(),
        }
        for f in files
    ]


@router.delete("/{file_id}")
async def delete_file(file_id: str, db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(UploadedFile).where(UploadedFile.id == file_id))
    f = result.scalar_one_or_none()
    if not f:
        raise HTTPException(status_code=404, detail="File not found")
    filepath = os.path.join(settings.upload_dir, f.filename)
    if os.path.exists(filepath):
        os.remove(filepath)
    await db.delete(f)
    await db.commit()
    return {"success": True}

File: cf-backend/src/routes/files.ts
// FILE: cf-backend/src/routes/files.ts
import { Hono } from "hono";
import { Env, autoSelectModel, genId } from "../types";
import { chatComplete } from "../services/ai";
import { uploadFile, deleteFile as deleteR2File, makeFileKey, getPublicUrl, extractTextFromFile } from "../services/storage";
import { saveFile, listFiles, getFile, deleteFile as deleteDbFile } from "../services/database";

export const filesRouter = new Hono<{ Bindings: Env }>();

const ALLOWED_TYPES = new Set([
  "pdf", "docx", "doc", "xlsx", "xls", "csv",
  "png", "jpg", "jpeg", "gif", "webp",
  "mp3", "wav", "m4a", "ogg", "webm", "mp4",
  "txt", "md", "json", "xml", "py", "js", "ts",
]);

filesRouter.post("/upload", async (c) => {
  const formData = await c.req.formData();
  const file = formData.get("file") as File | null;
  const question = (formData.get("question") as string) || "Analyze and summarize this file.";

  if (!file) return c.json({ error: "No file provided" }, 400);

  const ext = file.name.split(".").pop()?.toLowerCase() ?? "bin";
  if (!ALLOWED_TYPES.has(ext)) {
    return c.json({ error: `File type .${ext} not supported` }, 400);
  }

  const buffer = await file.arrayBuffer();
  const fileId = genId();
  const r2Key = makeFileKey(fileId, ext);

  // Upload to R2
  await uploadFile(c.env, r2Key, buffer, file.type || "application/octet-stream");

  // Extract text
  const extractedText = await extractTextFromFile(buffer, ext);

  // AI analysis
  const modelKey = autoSelectModel(question, true, ext);
  const prompt = `[File: ${file.name} | Type: ${ext.toUpperCase()} | Size: ${buffer.byteLength} bytes]

Extracted content:
${extractedText.slice(0, 5000)}

User question: ${question}`;

  const result = await chatComplete(
    c.env,
    [{ role: "user", content: prompt }],
    modelKey
  );

  // Save to D1
  const { id } = await saveFile(c.env, {
    originalName: file.name,
    fileType: ext,
    fileSize: buffer.byteLength,
    r2Key,
    contentSummary: result.text.slice(0, 1000),
    isKnowledgeBase: false,
  });

  return c.json({
    file_id: id,
    filename: file.name,
    file_type: ext,
    file_size: buffer.byteLength,
    parsed_text_length: extractedText.length,
    analysis: result.text,
    model_used: modelKey,
  });
});

filesRouter.get("/", async (c) => {
  const files = await listFiles(c.env);
  return c.json(
    files.map((f: any) => ({
      id: f.id,
      original_name: f.original_name,
      file_type: f.file_type,
      file_size: f.file_size,
      content_summary: f.content_summary,
      is_knowledge_base: Boolean(f.is_knowledge_base),
      created_at: new Date(f.created_at * 1000).toISOString(),
    }))
  );
});

// Serve file from R2
filesRouter.get("/download/:key{.+}", async (c) => {
  const key = decodeURIComponent(c.req.param("key"));
  const obj = await c.env.STORAGE.get(key);
  if (!obj) return c.json({ error: "File not found" }, 404);

  const headers = new Headers();
  obj.writeHttpMetadata(headers);
  headers.set("Cache-Control", "public, max-age=86400");

  return new Response(obj.body, { headers });
});

filesRouter.delete("/:id", async (c) => {
  const id = c.req.param("id");
  const file = await getFile(c.env, id);
  if (!file) return c.json({ error: "File not found" }, 404);

  await Promise.all([
    deleteR2File(c.env, (file as any).r2_key),
    deleteDbFile(c.env, id),
  ]);

  return c.json({ success: true });
});
